lib.simpleOptions: new function to simplify options declaration

hugosenari opened the pull request:
###### Description of changes

**Summary**

Declarative options function to make easier define new options.

**Motivation**

Inspired by Eelco Dolstra talk of NixCon 2020¹ and @DavHau pkgs-modules² project and my own user experince, I could say that modules are fun but define new options aren't. And maybe is the reason for some `extraOptions`, lack of typing in NixOS Modules³.

**Detailed design**

Define a new fluent interface for options definition where no function call is required for most common cases (type, listOf, attrsOf, options).

The logic is:
- if it has `attrsOf`, is attrOf
```nix
lib.simpleOptions {
  options.FOO.attrsOf  = types.str;
} == {
  options.FOO.type     = types.attrsOf types.str;
}
```
- if it has `listOf`, is listOf
 ```nix
lib.simpleOptions {
  options.BAR.listOf    = types.str;
} == {
  options.BAR.type      = types.listOf types.str;
}
```
- if it has `options`, is a submodule: 
```nix
lib.simpleOptions {
  options.EGGS.options.SPAM.type = types.str;
} == {
  options.EGGS.SPAM.type         = types.str;
} == {
  options.EGGS.type = submodule { options.SPAM.type = types.str; };
}
```
- or else `type` is required
```nix
lib.simpleOptions {
  options.TICO-TICO.type        = types.str;
} == {
  options.TICO-TICO.type        = types.str;
}
```

Create a helper function that translates this new (experimental) interface so no initial retro compatibility problem (except function not available) is created.

It looks not so different unless we add some nested example like options of attrs of options of list of string
```nix
lib.simpleOptions {
  options.FOO.attrsOf.options.BAR.listOf        = types.str;
} == {
  options.FOO.type = types.attrsOf ( types.submodule { options.BAR.type = types.listOf types.str; } );
}
```


**Example**:
This code (current way)

```nix
{ lib, ...}:
let
  Node = lib.types.submodule {
    options.tags = lib.mkOption {
      type        = lib.types.listOf lib.types.str;
      example     = ["logstash"];
      default     = [];
      description = "Define the tags of this node in the project cluster";
    };
  };
  env = lib.mkOption {
    type        = lib.types.attrsOf Node;
    default     = {};
    example     = { logstash01 = { tags = ["logstash"]; }; };
    description = "nodes of this environment";
  };

  Project = lib.types.submodule {
    options.git  = lib.mkOption {
      default     = "";
      type        = lib.types.str;
      example     = "https://github.com/DavHau";
      description = "git project url";
    };
    options.desc = lib.mkOption {
      default     = "";
      type        = lib.types.str;
      example     = "Elastic stack config files";
      description = "Description of the project";
    };
    options.docs = lib.mkOption {
      default     = "";
      type        = lib.types.str;
      example     = "https://google.com";
      description = "url of main documentations";
    };
    options.dev = env;
    options.prd = env;
  };
in
{
  options.project = lib.mkOption {
    type        = lib.types.attrsOf Project;
    default     = {};
    example     = { elk.desc = "elastic stack"; elk.dev.logstash01.tags = ["logstash"]; };
    description = "Information about our git projects";
  };
}
```

After proposed solution could be like this:

```nix
{lib, ...}:
let
  env.default      = {};
  env.example      = { logstash01 = { tags = ["logstash"]; }; };
  env.description  = "nodes of this environment";
  env.attrsOf.options.tags.listOf      = lib.types.str;
  env.attrsOf.options.tags.example     = ["logstash"];
  env.attrsOf.options.tags.default     = [];
  env.attrsOf.options.tags.description = "Define the tags of this node in the project cluster";
in
{
  options.project.default     = {};
  options.project.example     = { elk.desc = "elastic stack"; elk.dev.logstash01.tags = ["logstash"]; };
  options.project.description = "Information about our git projects";
  options.project.attrsOf.options = {
    git.default      = "";
    git.type         = lib.types.str;
    git.example      = "https://github.com/DavHau/pkgs-modules";
    git.description  = "git project url";
    desc.default     = "";
    desc.type        = lib.types.str;
    desc.example     = "Elastic stack config files";
    desc.description = "Description of the project";
    docs.default     = "";
    docs.type        = lib.types.str;
    docs.example     = "https://google.com";
    docs.description = "url of main documentations";
    dev = env;
    prd = env;
  };
}
```

**Drawbacks**

"I think the module system already has too much syntax sugar. Every time you add syntax sugar, you introduce at least one corner case." - @roberth 

"Furthermore, you obfuscate the original data model. You might flatten the very start of the learning curve, but a newcomer will have learned the wrong thing, and still needs to learn the original syntax before they become proficient." - @roberth

**Alternatives**

- Implement a [nickel-nix](https://github.com/nickel-lang/nickel-nix) modules translator.
- Create a stand alone project to mature the idea and documentation 

**Future work**

Should we have other types like `oneOf`, `nullOr`

**TODO**:

- [X] Configure tests
- [ ] Documentation
- [ ] Format code

¹. https://www.youtube.com/watch?v=dTd499Y31ig
². https://github.com/DavHau/pkgs-modules/issues/1
³. https://discourse.nixos.org/t/nixpkgs-nixos-is-an-expert-system-database/671/7



<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

edit: I edited it so many times that I lost the count :-D

hugosenari committed:
lib.simpleOptions: new function to simplify options declaration

**Summary**

Declarative options function to make easier define new options.

**Motivation**

Inspired by Eelco Dolstra talk of NixCon 2020¹ and @DavHau pkgs-modules²
project and my own user experince, I could say that modules are fun but define new
options aren't. And maybe is the reason for some `extraOptions`, lack of
typing in NixOS Modules³.

**Detailed design**

Define a new fluent interface for options definition where no function
call is required for most common cases (type, listOf, attrOf, options).

Create a helper function that translates this new (experimental)
interface so no initial retro compatibility problem (except function not
available) is created.

**Drawbacks**

"I think the module system already has too much syntax sugar. Every time you add syntax sugar, you introduce at least one corner case." - @roberth
"Furthermore, you obfuscate the original data model. You might flatten the very start of the learning curve, but a newcomer will have learned the wrong thing, and still needs to learn the original syntax before they become proficient." - @roberth

**Alternatives**

Implement a [nickel-nix](https://github.com/nickel-lang/nickel-nix)
modules translator.

**Future work**

Should we have other types like `oneOf`

**TODO**:

- [ ] Configure tests (we have teste cases but aren't configured)
- [ ] Documentation
- [ ] Format code

¹. https://www.youtube.com/watch?v=dTd499Y31ig
². https://github.com/DavHau/pkgs-modules/issues/1
³. https://discourse.nixos.org/t/nixpkgs-nixos-is-an-expert-system-database/671/7

hugosenari committed:
lib.simple-options: fix type in default

roberth commented on lib/simple-options.nix:
```diff
@@ -0,0 +1,89 @@
+      if builtins.hasAttr typeAttr optDef
+        then optDef else
+      if builtins.hasAttr attrAttr optDef
+        then builtins.removeAttrs optDef [attrAttr] // { type = lib.types.attrsOf   (toType    (breadcrumb ++ [optName attrAttr]) optDef."${attrAttr}");} else
```
I would recommend `lazyAttrsOf`, which supports more definitions without causing infinite recursion. The only reason to use `attrsOf` is `mkIf`, but `mkIf` does not really add anything to an `attrsOf` definition, as it is no lazier than `optionalAttrs` would be in this case, iirc. It is not used as often as `attrsOf` though, because in many cases it isn't necessary, and even if it is, people just give up because the infinite recursion is hard to debug.

roberth commented on lib/simple-options.nix:
```diff
@@ -0,0 +1,89 @@
+        then builtins.throw ''breadcrumb must be a list, instead it is a ${builtins.typeOf breadcrumb}'' else
+      if builtins.typeOf type       != "set"
+        then builtins.throw ''type must be a attrset, instead it is a ${builtins.typeOf type}, breadcrumb: ${builtins.concatStringsSep "." breadcrumb}'' else
+      if builtins.hasAttr "_type" type
+        then type else
```
```suggestion
      type._type or (
```

This is also a micro-optimization: fewer function calls.
I have to admit that I haven't measured it though.

roberth reviewed (commented):
Not eager to merge this yet, as we'll have to teach everyone and our documentation is already strained, but I think it's worth playing around with.

hugosenari committed:
lib.simpleOptions: test new function to simplify options declaration

hugosenari committed:
lib.simpleOptions: test new function to simplify options declaration

hugosenari committed:
lib.simpleOptions: function to simplify options declaration

> A bit more readable, plus this saves some . lookups where you would otherwise have to write (builtins.), leaving only the double array index that would otherwise be spent on builtins anyway.
> Extend the list of identifiers to remove the other lookups as well.
by @roberth

hugosenari committed:
lib.simpleOptions: function to simplify options declaration

Remove hasAttr calls

> This is also a micro-optimization: fewer function calls.
> I find this more readable, and it avoids function calls.

by @roberth

hugosenari committed:
lib.simpleOptions: function to simplify options declaration

Change attrsOf by lazyAttrsOf

> I would recommend lazyAttrsOf, which supports more definitions without causing infinite recursion. The only reason to use attrsOf is mkIf, but mkIf does not really add anything to an attrsOf definition, as it is no lazier than optionalAttrs would be in this case, iirc. It is not used as often as attrsOf though, because in many cases it is not necessary, and even if it is, people just give up because the infinite recursion is hard to debug.

by @roberth

hugosenari commented on lib/simple-options.nix:
```diff
@@ -0,0 +1,91 @@
+      if type ? _type
+        then type else
+      if type ? ${typeAttr}
+        then toType type.type else
```
```suggestion
        then toType (breadcrumb ++ [typeAttr]) type.${typeAttr} else
```

hugosenari commented on lib/simple-options.nix:
```diff
@@ -0,0 +1,91 @@
+      if typeOf optDef     != "set"
+        then throw ''optDef must be a attrset, instead it is a ${typeOf optDef}, breadcrumb: ${concatStringsSep "." breadcrumb}, optName: ${optName}'' else
+      if optDef ? ${typeAttr}
+        then optDef else
```
```suggestion
        then removeAttrs optDef [typeAttr] // { type = optDef.${typeAttr}; } else    
```

hugosenari commented on lib/simple-options.nix:
```diff
@@ -0,0 +1,91 @@
+      if optDef ? ${typeAttr}
+        then optDef else
+      if optDef ? ${attrAttr}
+        then removeAttrs optDef [attrAttr] // { type = lazyAttrsOf (toType    (breadcrumb ++ [optName attrAttr]) optDef."${attrAttr}");} else
+      if optDef ? ${listAttr}
+        then removeAttrs optDef [listAttr] // { type = listOf      (toType    (breadcrumb ++ [optName listAttr]) optDef."${listAttr}");} else
+      if optDef ? ${subMAttr}
+        then removeAttrs optDef [subMAttr] // { type = submodule   (toOptions (breadcrumb ++ [optName subMAttr]) optDef."${subMAttr}");}
```
```suggestion
        then removeAttrs optDef [attrAttr] // { type = lazyAttrsOf (toType    (breadcrumb ++ [optName attrAttr]) optDef.${attrAttr});} else
      if optDef ? ${listAttr}
        then removeAttrs optDef [listAttr] // { type = listOf      (toType    (breadcrumb ++ [optName listAttr]) optDef.${listAttr});} else
      if optDef ? ${subMAttr}
        then removeAttrs optDef [subMAttr] // { type = submodule   (toOptions (breadcrumb ++ [optName subMAttr]) optDef.${subMAttr});}
```

hugosenari committed:
lib.simpleOptions: function to simplify options declaration

fix type.{typeAttr} and otpDfe.{typeAttr} where typeAttr != "type" bug

hugosenari commented:
> Not eager to merge this yet, as we'll have to teach everyone and our documentation is already strained, 

Ok, adding "stand alone project" to PR description "alternatives" section.

> but I think it's worth playing around with.

Check cruel-intentions/nixpkgs#1 for some test changing services/misc/nix-daemon.nix and system/boot/systemd.nix to use this function

And cruel-intentions/nixpkgs#2 for nullOr (alias), mdDoc (alias) and type inference of default

hugosenari commented:
I was thinking, the best use case for this is complex options (mix of attr/list/submodule) instead of simple options, maybe I give the function a bad name :thinking:

hugosenari commented:
> Ok, adding "stand alone project" to PR description "alternatives" section.

["Stand alone project"](https://github.com/cruel-intentions/flu-type-a) to use it in another projects until we have proper documentation/consensus, also helps validate the idea.

hugosenari commented on lib/simple-options.nix:
```diff
@@ -0,0 +1,89 @@
+  ...
+}:
+module:
+let
```
Related meme.

![image](https://github.com/NixOS/nixpkgs/assets/863299/53e82e8a-2142-46da-a143-68bcca51e2e1)
